### ISSUE TITLE
Unable to load custom hetero dataset with different edge triplets

### ISSUE THREAD

**Issue by pedrofrancescon** [pyg-team/pytorch_geometric#3984]
### 🐛 Describe the bug

I'm building a custom dataset with multiple heterogeneous graphs. When trying to access some individual graphs by the dataset indices I get the following error:

```bash
Traceback (most recent call last):
  File "/home/.../graph_init.py", line 199, in <module>
    print(dataset[1])
  File "/home/.../anaconda3/envs/pyg/lib/python3.9/site-packages/torch_geometric/data/dataset.py", line 198, in __getitem__
    data = self.get(self.indices()[idx])
  File "/home/.../anaconda3/envs/pyg/lib/python3.9/site-packages/torch_geometric/data/in_memory_dataset.py", line 91, in get
    data = separate(
  File "/home/.../anaconda3/envs/pyg/lib/python3.9/site-packages/torch_geometric/data/separate.py", line 41, in separate
    data_store[attr] = _separate(attr, batch_store[attr], idx, slices,
  File "/home/.../anaconda3/envs/pyg/lib/python3.9/site-packages/torch_geometric/data/separate.py", line 90, in _separate
    start, end = slices[idx], slices[idx + 1]
IndexError: index 2 is out of bounds for dimension 0 with size 2
```

For this dataset I have some graphs that don't have a specific edge type (triplet), as such: 

```bash
HeteroData(
  V={ x=[18901, 2] },
  C={ x=[2053, 2] },
  P={
    x=[2620, 97],
    y=[2620]
  },
  [...]
  (P, data, C)={
    edge_index=[2, 2641],
    edge_attr=[2641]
  },
  (P, data, V)={
    edge_index=[2, 1],
    edge_attr=[1]
  }
)
HeteroData(
  V={ x=[6336, 2] },
  C={ x=[782, 2] },
  P={
    x=[867, 97],
    y=[867]
  },
  [...]
  (P, data, C)={
    edge_index=[2, 867],
    edge_attr=[867]
  }
)
[...]
```
So the `self.slices` object when loading the dataset looks like this:

```bash
defaultdict(<class 'dict'>, {
'V': {'x': tensor([    0, 18901, 25237, 32134, 39745])}, 
'C': {'x': tensor([   0, 2053, 2835, 3683, 4586])}, 
'P': {'x': tensor([   0, 2620, 3487, 4438, 5392]), 'y': tensor([   0, 2620, 3487, 4438, 5392]), 
[...]
('P', 'data', 'C'): {'edge_index': tensor([   0, 2641, 3508, 4459, 5413]), 'edge_attr': tensor([   0, 2641, 3508, 4459, 5413])}, 
('P', 'data', 'V'): {'edge_index': tensor([0, 1]), 'edge_attr': tensor([0, 1])}})
```
The dataset currently has 4 graphs, but the size of the last tensor on the `slices` object doesn't match with it, making it impossible to load any other than the first graph.

### Environment

* PyG version: 2.0.2
* PyTorch version: 1.10.0
* OS: Linux (Kubuntu)
* Python version: 3.9.7
* CUDA/cuDNN version: 11.6
* How I installed PyTorch and PyG: conda

**Comment by rusty1s** [comment]
I see. Currently, we expect that all data that you collate together have the same attributes and types. You can easily fix that by adding some dummy attributes, e.g.:
```python
data['P', 'data', 'C'].edge_index = torch.empty(2, 0, dtype=torch.long)
```
I will also look into "fixing" that on PyG end as well.

**Comment by pedrofrancescon** [comment]
Awesome, thank you for the hint as well!

**Comment by alvarogunawan** [comment]
I'm using a similar custom dataset where certain graphs that lack a certain edge type and as such have added dummy attributes for the missing edges. However, I find that when using `to_hetero()` and passing the data containing the dummy attributes, I run into the following error:

```
Traceback (most recent call last):
  File "/home/em11824/.local/lib/python3.8/site-packages/torch/fx/graph_module.py", line 608, in wrapped_call
    return super(type(self), self).__call__(*args, **kwargs)
  File "/home/em11824/.local/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "<eval_with_key>.1", line 1157, in forward
    hidden_gat__disj34 = torch.add(hidden_gat__disj1, hidden_gat__disj2);  hidden_gat__disj1 = hidden_gat__disj2 = None
RuntimeError: The size of tensor a (374) must match the size of tensor b (7) at non-singleton dimension 0

Call using an FX-traced Module, line 1157 of the traced Module's generated forward function:
    hidden_gat__neg = torch.add(hidden_gat__neg63, hidden_gat__neg64);  hidden_gat__neg63 = hidden_gat__neg64 = None
    hidden_gat__disj34 = torch.add(hidden_gat__disj1, hidden_gat__disj2);  hidden_gat__disj1 = hidden_gat__disj2 = None

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    hidden_gat__disj35 = torch.add(hidden_gat__disj3, hidden_gat__disj4);  hidden_gat__disj3 = hidden_gat__disj4 = None

    hidden_gat__disj36 = torch.add(hidden_gat__disj5, hidden_gat__disj6);  hidden_gat__disj5 = hidden_gat__disj6 = None

Traceback (most recent call last):
  File "model_graph_2.py", line 627, in <module>
    main()
  File "model_graph_2.py", line 93, in main
    out = model(data.x_dict, data.edge_index_dict)
  File "/home/em11824/.local/lib/python3.8/site-packages/torch/fx/graph_module.py", line 616, in wrapped_call
    raise e.with_traceback(None)
RuntimeError: The size of tensor a (374) must match the size of tensor b (7) at non-singleton dimension 0

```

**Comment by rusty1s** [comment]
Do you have a small example to reproduce? This would be of great help!

**Comment by alvarogunawan** [comment]
Sorry for the late response. Following some advice from #4143, I attempted to reproduce this issue with `FakeHeteroDataset()` and `to_hetero()` (also including an empty node type). The following example reproduces the previous error I was encountering with my dataset:

```python
import torch
from torch_geometric.nn import Linear, GATConv, to_hetero
from torch_geometric.datasets import FakeHeteroDataset

class HeteroNet(torch.nn.Module):
    def __init__(self):
        super(HeteroNet, self).__init__()
        self.fc_1_input = Linear(-1, 64)

        self.hidden_gat = GATConv((-1,-1), 64, heads=4, add_self_loops=False)

        self.fc_1_output = Linear(-1, 1)
            

    def forward(self, x, edge_index):
        x = self.fc_1_input(x)
        x = self.hidden_gat(x, edge_index)
        x = self.fc_1_output(x)
        return x
    
        
data = FakeHeteroDataset().data
data["extra"].x = torch.empty(0, 10, dtype=torch.float)

data['v0', 'e0', 'extra'].edge_index = torch.empty(2, 0, dtype=torch.long) 
data['extra', 'e0', 'extra'].edge_index = torch.empty(2, 0, dtype=torch.long) 

model = HeteroNet()
model = to_hetero(model, data.metadata())
out = model(data.x_dict, data.edge_index_dict)
print(out)
```
This generates the error:

```shell
  File "weird_test.py", line 42, in <module>
    out = model(data.x_dict, data.edge_index_dict)
  File "/home/em11824/.local/lib/python3.8/site-packages/torch/fx/graph_module.py", line 616, in wrapped_call
    raise e.with_traceback(None)
RuntimeError: The size of tensor a (935) must match the size of tensor b (0) at non-singleton dimension 0
```

There's some other oddities I found: for some reason, changing what empty edge types are present causes different errors. I've annotated some of the combinations below:

```python
data = FakeHeteroDataset().data
data["extra"].x = torch.empty(0, 10, dtype=torch.float)

# adding only this generates no errors
data['v0', 'e0', 'extra'].edge_index = torch.empty(2, 0, dtype=torch.long) 

# adding only this generates no errors
data['extra', 'e0', 'extra'].edge_index = torch.empty(2, 0, dtype=torch.long)

# adding both of these introduces the RuntimeError seen above
data['v0', 'e0', 'extra'].edge_index = torch.empty(2, 0, dtype=torch.long) 
data['extra', 'e0', 'extra'].edge_index = torch.empty(2, 0, dtype=torch.long) 

# adding only this introduces "AttributeError: 'NoneType' object has no attribute 'size'" error, seen below
data['extra', 'e0', 'v0'].edge_index = torch.empty(2, 0, dtype=torch.long) 

# adding this pair together generates no errors
data['v0', 'e0', 'extra'].edge_index = torch.empty(2, 0, dtype=torch.long)
data['extra', 'e0', 'v0'].edge_index = torch.empty(2, 0, dtype=torch.long)

# adding this pair together also generates no errors
data['extra', 'e0', 'extra'].edge_index = torch.empty(2, 0, dtype=torch.long) 
data['extra', 'e0', 'v0'].edge_index = torch.empty(2, 0, dtype=torch.long) 
```

The other error mentioned above is the following:

```shell
Traceback (most recent call last):
  File "weird_test.py", line 28, in <module>
    out = model(data.x_dict, data.edge_index_dict)
  File "/home/em11824/.local/lib/python3.8/site-packages/torch/fx/graph_module.py", line 616, in wrapped_call
    raise e.with_traceback(None)
AttributeError: 'NoneType' object has no attribute 'size'

```

**Comment by rusty1s** [comment]
Thanks for the example. This was really helpful. I fixed it in #4222. This was a minor but stupid bug, sorry :(
